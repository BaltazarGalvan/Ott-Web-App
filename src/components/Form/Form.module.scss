@use '../../styles/variables';
@use '../../styles/theme';
@use '../../styles/mixins/responsive';
