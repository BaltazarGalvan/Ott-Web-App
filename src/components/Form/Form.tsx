import React, { useState } from 'react';

// import styles from './Form.module.scss';

type FormValues = Record<string, string>;

type Return = {
  values: FormValues;
  onChange: (event: React.FormEvent<HTMLInputElement>) => void;
  handleSubmit: () => void;
};

type Props = {
  initialValues: FormValues;
  children: ({ values, onChange }: Return) => JSX.Element;
  onSubmit: (values: FormValues) => void;
};

const Form = ({ initialValues, children, onSubmit }: Props): JSX.Element => {
  const [values, setValues] = useState<FormValues>(initialValues);

  const onChange = (event: React.FormEvent<HTMLInputElement>) => {
    if (!event.currentTarget) return;
    setValues({ ...values, [event.currentTarget.name]: event.currentTarget.value });
  };

  const handleSubmit = (event?: React.FormEvent) => {
    event && event.preventDefault();
    onSubmit(values);
  };

  return <form onSubmit={handleSubmit}>{children({ values, onChange, handleSubmit })}</form>;
};

export default Form;
