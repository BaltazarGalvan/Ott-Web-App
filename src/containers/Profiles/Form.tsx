import { useEffect } from 'react';
import { type SchemaOf, object, string } from 'yup';
import { useNavigate } from 'react-router';

import profileStyles from './Profiles.module.scss';

import Button from '#src/components/Button/Button';
import Dropdown from '#src/components/Dropdown/Dropdown';
import FormFeedback from '#src/components/FormFeedback/FormFeedback';
import TextField from '#src/components/TextField/TextField';
import useForm, { UseFormOnSubmitHandler } from '#src/hooks/useForm';
import styles from '#src/pages/User/User.module.scss';
import type { ProfilePayload } from '#types/account';
import LoadingOverlay from '#src/components/LoadingOverlay/LoadingOverlay';

type Props = {
  initialValues: ProfilePayload;
  formHandler: UseFormOnSubmitHandler<ProfilePayload>;
  setFullName?: (name: string) => void;
};

const Form = ({ initialValues, formHandler, setFullName }: Props) => {
  const navigate = useNavigate();

  const options: { value: string; label: string }[] = [
    { label: 'Adult', value: 'true' },
    { label: 'Kids', value: 'false' },
  ];

  const validationSchema: SchemaOf<{ name: string }> = object().shape({
    name: string().required('You must enter your full name').min(2, 'You must include at least 2 characters'),
  });

  const { handleSubmit, handleChange, values, errors, submitting } = useForm(initialValues, formHandler, validationSchema);

  useEffect(() => {
    setFullName?.(values.name);
  }, [values, setFullName]);

  const formLabel = values?.id ? 'Edit your profile' : 'Create your profile';

  return (
    <form onSubmit={handleSubmit} noValidate>
      <div className={styles.panel}>
        <h2 className={styles.panelHeader}>{formLabel}</h2>
        <div className={profileStyles.formFields}>
          {errors.form ? <FormFeedback variant="error">{errors.form}</FormFeedback> : null}
          {submitting && <LoadingOverlay inline />}
          <TextField
            required
            name="name"
            label="Full name"
            value={values?.name}
            onChange={handleChange}
            error={!!errors.name || !!errors.form}
            helperText={errors.name}
          />
          <Dropdown
            fullWidth
            required
            name="adult"
            label="Content rating"
            helperText="Content 12+ will not be available for kids profiles"
            className={styles.dropdown}
            options={options}
            value={values?.adult?.toString() || 'true'}
            onChange={handleChange}
          />
        </div>
        <>
          <Button type="submit" label="Save" variant="outlined" disabled={submitting} />
          <Button onClick={() => navigate('/u/profiles')} label="Cancel" variant="text" />
        </>
      </div>
    </form>
  );
};

export default Form;
