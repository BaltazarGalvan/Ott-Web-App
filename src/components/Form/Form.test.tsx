import * as React from 'react';
import { render } from '@testing-library/react';

import Form from './Form';

describe('<Form>', () => {
  const initialValues = { test: 'Testing' };

  it('renders Form', () => {
    const { container } = render(
      <Form initialValues={initialValues} onSubmit={() => null}>
        {({ values, onChange }) => <input name="test" value={values.test || ''} onChange={onChange} />}
      </Form>,
    );
    expect(container).toMatchSnapshot();
  });
});
