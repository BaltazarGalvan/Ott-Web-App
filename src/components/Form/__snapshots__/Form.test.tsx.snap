// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form> renders Form 1`] = `
<div>
  <form>
    <input
      name="test"
      value="Testing"
    />
  </form>
</div>
`;
